flux-middle/src/rty/fold.rs

```diff
@@ -195,6 +195,32 @@ pub trait TypeSuperVisitable: TypeVisitable {
     fn super_visit_with<V: TypeVisitor>(&self, visitor: &mut V) -> ControlFlow<V::BreakTy, ()>;
 }
 
+#[derive(Debug)]
+struct ProjectionTable(FxHashMap<AliasTy, Ty>);
+
+impl ProjectionTable {
+    pub fn new(predicates: GenericPredicates) -> Self {
+        let mut res = FxHashMap::default();
+        for pred in &predicates.predicates {
+            if pred.kind.vars().is_empty() {
+                if let ClauseKind::Projection(proj_pred) = pred.kind.clone().skip_binder() {
+                    // TODO(RJ): duplicate entries overwrite?
+                    res.insert(proj_pred.projection_ty, proj_pred.term);
```

Comment:
Rust should catch some of the ambiguities if they are multiple matching clauses, but there's also a notion of precedence, so to be on the safe side let's just panic here and we will learn about this rule in due course.